### PR TITLE
Provide user more information during ping discovery flow. Closes #302. Closes #326.

### LINKED ISSUE #302: If one host in profile with multiple hosts is unreachable, no error or message is generated
## Specify type:
- Bug

### Bug severity (if applicable):
- Medium
___

## Description:
If a profile is scanned that contains at least one valid host, the scan completes with no complaints, even if there were unreachable hosts in the profile.
___

### Version of rho:
```
version: 0.0.28
	ansible: 2.3.2.0
	python: 3.6.2 (default, Jul 24 2017, 14:22:34) 
```

### Expected behavior:
While I think it is good that rho goes on and scans the reachable systems, I think it should say *something* about the fact there were unreachable hosts.

This can be a hair-pulling bug if you are scanning lots of systems and are trying to figure out why your report contains less rows than you have hosts in the profile.

### Actual behavior:
If at least one host is reachable, the scan completes with no complaints and generates a report file with only information pertaining to the reachable hosts.

### Steps to reproduce:
0) Create an auth for a system you have access to and is alive
1) Create a profile with at least one valid host and one invalid host and the auth that will grant you access to the valid host.
2) Run a scan against that profile.

### Environment information:
Provide detailed information on the *scanning* and *target* systems (if applicable); see example below.

Type            | Operating System | OS Version | Python Version | Virtualization
--------------- | ---------------- | ---------- | -------------- | --------------
Scanning System | Fedora             | 25        | 3.6            | Bare Metal
Target System   | RHEL             | 7.4        | 2.7.5          | ESXi
Target System   | Invalid Host             | --       | --          | --

### LINKED ISSUE #326: No output is shown during the ping phase
# Openning a new issue
___

## Specify type:
- Enhancement

### Bug severity (if applicable):
- Medium
___

## Description:
No results are shown from the ping phase when mapping auths to hosts. I think we should at least show a status message letting the user know that's what's happening because in a large environment this could take a long time and with no output the user might believe the script is hung. Maybe a similar approach to the regular output where it is dependent on the verbosity level but it at least always shows something to let the user know work is being done. 
___

### PR DESCRIPTION
**Pull request by chambridge**
Scans without verbosity will look as follows:
```
$ rho scan --profile testenv --reportfile foo.csv 
Please enter your rho vault password: 
Connection discovery will be perform with the following auth credentials: master
Attempting connection discovery with auth "master".

Connection succeeded with auth "master" to 3 systems.
Failed to connect with auth "master" to 19 systems.

Failed to connect to 19 systems with all auth credentials. See the following file "/var/folders/tr/w6c0zbvs0jzd25027nyyw06r0000gn/T/tmp173Jug" for a list  of the failed systems.

Scan will be performed against 3 of 22 systems.
```

Scans with verbosity > 1 will show the failed system output aswell:
```
$ rho scan --profile testenv --reportfile foo.csv -vv
Please enter your rho vault password: 
Connection discovery will be perform with the following auth credentials: master
Attempting connection discovery with auth "master".

Connection succeeded with auth "master" to 3 systems.
Failed to connect with auth "master" to 19 systems.

Failed to connect to 19 systems with all auth credentials. See the following file "/var/folders/tr/w6c0zbvs0jzd25027nyyw06r0000gn/T/tmpHqlViM" for a list  of the failed systems.
Failed to connect to the following systems: 10.10.182.110, 10.10.181.253, 10.10.182.244, 10.10.181.192, 10.10.182.50, 10.10.181.50, 10.10.182.250, 10.10.182.201, 10.10.181.9, 10.10.182.95, 10.10.181.132, 10.10.181.221, 10.10.181.205, 10.10.181.3, 10.10.181.209, 10.10.181.154, 10.10.182.194, 10.10.181.14, 10.10.181.72.

Scan will be performed against 3 of 22 systems.
```

Also as a note the temp file created with the list of failed hosts is new line delimited so it could be consumed back into a new profile with the `--hosts` command if the user decided to do so.

### CODE REVIEW

**Comment by elyezer** [comment]
This is a good improvement but there is one potential issue with this output, if the user have ssh-agent running and have a valid ssh key rho will report as success while it should be a failure.

Try this:

* create an auth credential with a SSH key which is not valid for a host on the host list, but the key on your ssh-agent is
* create a profile with those hosts
* run scan on those hosts and check the output, it will likely show that it was able to connect

With that user will see that it was possible to connect even though it should not. So this output is for sure informative but may led to incorrect information because Ansible delegate the connection to the ssh client if it is available, when it is not available then it will use Paramiko and that will properly state a failure.

**Comment by chambridge** [comment]
@elyezer What is your suggestion to do around this? I get that there could be a disconnect between us saying that it connects with credential "xyz" and it actually used some other credentials.

Would you prefer I add some output note around pre-configured ssh-agent credentials? Should we add this into the man file somewhere?

**Comment by elyezer** [comment]
Even though I like having on manual more it seems that not everyone reads it so I would say that maybe both places would be good to have this information.

I mean this is an edge case but may led to some bug reports from the users.

**Comment by mdvickst** [comment]
Is this really a bad thing? The goal is to discover more systems so if the did not specify a auth that works but the ssh connection is successful due to their SSH agent that's a good thing. I'm just not sure how to document the user/auth it worked with.

**Comment by chambridge** [comment]
@mdvickst  I think @elyezer was just hoping to provide some documentation as the output suggest the connectivity was gained with a specific credential. I've added a small note to the output and man page.

File: doc/source/man.rst [hunk 1]
```diff
@@ -281,7 +281,8 @@ To encrypt sensitive facts within the comma separated values (CSV) file output o
 Scanning
 --------
 
-Use the ``rho scan`` command to run discovery and inspection scans on the network. This command scans all of the host names or IP addresses that are defined in the supplied network profile, and then writes the report information to a comma separated values (CSV) file.
+Use the ``rho scan`` command to run discovery and inspection scans on the network. This command scans all of the host names or IP addresses that are defined in the supplied network profile, and then writes the report information to a comma separated values (CSV) file. Note: Any ssh-agent connection setup for a target host '
+              'will be used as a fallback if it exists.
```
Comment (noahl) [inline doc/source/man.rst:285]: Random question: the ssh-agent thing might actually be a feature for people who have ssh-agent already set up. Should we allow profiles with zero auths as a convenience for those people?

**Review by noahl — APPROVED** [review]

**Pull request merged**